visual-crowd-preview Pipeline with Fieldtrip

Simo149 opened the pull request:
adding the MEG analysis code that i am working on, as well as the stimuli set.

Simo149 commented on experiments/psychtoolbox/visual_crowding_preview/CP_main.m:
I also uploaded the .con, .mrk and .mat files to the nyu box.

Hzaatiti commented on experiments/psychtoolbox/visual_crowding_preview/meg_analysis.m:
```diff
@@ -0,0 +1,71 @@
+% Notch filter the data at 50 Hz
+cfg = [];
+cfg.bsfilter = 'yes';
+cfg.bsfreq = [49 51]; % Notch filter range
```
same, this should be usually corrected by the baseline correction step, but comment it for now and w'll see its effect on the results later

Hzaatiti commented on experiments/psychtoolbox/visual_crowding_preview/meg_analysis.m:
```diff
@@ -0,0 +1,71 @@
+
+time = data_filtered_EC.time{1}; % Extract time vector from struct
+
+% plot raw and filtered data
```
Please add comments saying your intention from the next code block

Example: Now we will filter the data by a bandpass in order to reduce the noise

Hzaatiti replied:
Please divide your pipeline into code blocks using %%

Hzaatiti commented on experiments/psychtoolbox/visual_crowding_preview/meg_analysis.m:
```diff
@@ -0,0 +1,71 @@
+
+% Band-pass filter the data 
+cfg = [];
+cfg.bpfilter = 'yes';
```
even later when u need to use this filter, u will apply it only on the trials data and not the whole data, so this section will be moved to a later step in your code

Hzaatiti reviewed (changes requested):
please address the comments, and start defining your trials
define how many types of trials you need
and use the script to retrieve the triggers on all the channels you used https://github.com/hzaatiti-NYU/meg-pipeline/blob/main/pipeline/field_trip_pipelines/oddball_pipeline_beamformer/KIT/readfif_ft.m

When this is done, we can start defining the trials similarly as in 
https://github.com/hzaatiti-NYU/meg-pipeline/blob/main/pipeline/field_trip_pipelines/oddball_pipeline_beamformer/KIT/oddball_kit_MEG_clean_sensor.m

Simo149 commented:
I segmented the data based on the preview event (channel 227). I have also tried to Inspect and reject trials using the function ft_rejectvisual, however I couldn't browse through the trials as described in the fieldtrip tutorial. I could only see trial 1. 
I added a code block to separate trials based on the crowding conditions in my experiment but haven't run this section yet (commented out). 
I also kept the filtered data for the time being so I can inspect the differnce between the filtered data and the raw data when using the ft_rejectvisual function.

hzaatiti-NYU commented:
Hi Tas your notebook is available here!
https://meg-pipeline.readthedocs.io/en/latest/5-pipeline/notebooks/fieldtrip/fieldtrip_visual_crowd_preview_pipeline.html
Let's fill it up with what you have already so far and continue building the ATLAS part.
To execute the notebook locally you will need to install jupyter integration with matlab https://github.com/mathworks/jupyter-matlab-proxy

Simo149 commented:
The complete preprocessing script is now in the file 'MEG_analysis_ERP.m' and I have removed the file 'meg_analysis.m'. I also added the scripts for averaging over all subjects  and co-registration. The code execution order should be in the following order:
1. 'MEG_analysis_ERP.m'
2. 'avgSubjects.m'
3. 'coregistration.m'

It is also important to note that I have a different file order structure for the subjects' data, as it was easier to loop over subjects this way. The structure is as follows:

LaserScan Data\Sub_00*_scan_vcp.txt
Markers\Sub_00*_1_vcp.mrk
MatLab Data\Sub_00*_vcp.mat
MEG Data\Sub_00*_01_vcp.con
This file structure allowed for easier loop coding.

NOTE: ERP results are still not accurate due to an error in the .con file

Hzaatiti commented:
add a counter structure for each channel to count the flips

Hzaatiti commented on experiments/psychtoolbox/visual_crowding_preview/README.md:
```diff
@@ -1,13 +1,37 @@
+The stimuli were presented as either valid or invalid, with the invalid stimulus being a flipped version of the valid one.
+After that, the participant has to decide weather or not the word displayed is the same your they saw during the trial by answering the question 'Is this the last word you saw?'.
+The study employs magnetoencephalography (MEG) and eye-tracking techniques to record brain activity.
+
```
get me the references of the two papers chinese english on visual crowding

Hzaatiti commented:
@Simo149 
- add saving variables to avoid recomputation everytime we change parameters value
- follow the coregistration steps as described in https://meg-pipeline.readthedocs.io/en/latest/5-pipeline/notebooks/fieldtrip/fieldtrip_oddball_KIT_pipeline.html   and report back any problems or wrong result (coregistration not showing correct alignment)
- When doing beamformer without contrasting conditions, you need to correct for the beamformer bias 
To correct, more information here: https://www.fieldtriptoolbox.org/tutorial/beamformer/ and cell 15 in https://meg-pipeline.readthedocs.io/en/latest/5-pipeline/notebooks/fieldtrip/fieldtrip_oddball_OPM_pipeline.html#Beamformer-source-localization

contsili reviewed (commented):
These are my suggestions on your single-subject ERP analysis. Feel free to ask me whatever you need. Next week, I will move on with source reconstruction.

Simo149 commented on pipeline/field_trip_pipelines/visual_crowd_preview_pipeline/MEG_analysis_ERP.m:
```diff
@@ -0,0 +1,507 @@
+        % Detect transitions from low to high
+        transitions = diff(previewTrigger > threshold);
+        difference = previewTrigger >threshold;
+        counts = countSequencesWithAtLeastTwoOnes(transitions);
```
this was added as an additional sanity check. It can be commented out and will not affect the analysis.

contsili commented on pipeline/field_trip_pipelines/visual_crowd_preview_pipeline/coregistration.m:
```diff
@@ -0,0 +1,143 @@
+% p1t = ft_warp_apply(transform_mrk2laser, p1)
+
+grad = ft_read_sens(confile,'senstype','meg');
+grad= ft_transform_geometry(transform_mrk2laser, grad);
```
I plotted the MEG sensors and the headshape and they are not aligned correctly with your method. Instead, you should use ft_electroderealign. I will explain to you tomorrow in our online meeting.

contsili commented on pipeline/field_trip_pipelines/visual_crowd_preview_pipeline/coregistration.m:
```diff
@@ -0,0 +1,143 @@
+
+mrk1 = ft_read_headshape(mrkfile1);
+mrk1 = ft_convert_units(mrk1, lasershape.unit);
+mrk2 = ft_read_headshape(mrkfile2);
```
Ans. mrk2 are the positions of the coils at the end of the experiment

contsili commented:
The **single-subject analysis** is complete.

Since we have no individual mri we need to use a template mri or headmodel:

1. `beamformer_mri_KT.m`: use a template mri (in MNI space) and aligns it with the polhemus  (in subject space)
2. `beamformer_headmodel_KT.m`: use a template headmodel (in MNI space) and aligns it with the polhemus (in subject space)

I recommend:`beamformer_mri_KT.m` as it makes group analysis easier later, and has better visualizations of our results.


Once you know based on your research question what conditions you are going to contrast and at what latencies, you can adjust the code to answer your research question. 

For example, I looked 260-290 ms for the high crowding condition, and these are the results I get:
![sensor-source](https://github.com/user-attachments/assets/1ca04f38-2271-4b55-bfb2-8dc8b6829f94)

1s plot: red - no crowding, green - medium crowding, blue - high crowding
2nd and 3rd plo: the sensor and source topography of the high crowding condition

Next week I will check the **group analysis**: averaging in sensors and source space. After that everything should be in place to finish your thesis.

Feel free to ask me if you need something :)

Hzaatiti commented:
can we please sync this branch with the main branch? thank you

contsili commented:
> can we please sync this branch with the main branch? thank you

we cannot merge this branch to [Hzaatiti:main](https://github.com/Hzaatiti/meg-pipeline/tree/main) since two of the GH actions do not pass the test. What are you suggesting to do?

Hzaatiti commented:
> > can we please sync this branch with the main branch? thank you
> 
> we cannot merge this branch to [Hzaatiti:main](https://github.com/Hzaatiti/meg-pipeline/tree/main) since two of the GH actions do not pass the test. What are you suggesting to do?

By sync, I mean pull the changes from the main to this branch and solve any conflicts, because the tests have been changed, so they should pass after pulling the changes from main

contsili commented:
@Hzaatiti done

contsili commented:
We can't merge since some of your reviews have not been resolved. Shall we wait for @Simo149 to have a look at them and resolve them herself?

Hzaatiti commented:
> We can't merge since some of your reviews have not been resolved. Shall we wait for @Simo149 to have a look at them and resolve them herself?

no problem for that, i was just checking that the tests pass now and they do!
(the two new actions generate pdf report from the sphinx docs ^_^)

Hzaatiti commented:
@Simo149   to merge this PR we will need you to resolve the questions though, please mark them as resolved

contsili commented:
I also added the **group analysis** in sensor and source space. Now we have a complete analysis.

I also added the **statistics** based o the [cluster-based permutation framework](https://www.fieldtriptoolbox.org/tutorial/cluster_permutation_timelock/). @Simo149 Based on your research questions, you should adjust these statistics.
